mlx/backend/metal/metal.cpp

```diff
@@ -55,17 +55,19 @@ inline void check_error(MTL::CommandBuffer* cbuf) {
   }
 }
 
-std::function<void()> make_task(
-    array& arr,
-    std::vector<std::shared_future<void>> deps,
-    std::shared_ptr<std::promise<void>> p) {
-  auto task = [arr, deps = std::move(deps), p = std::move(p)]() mutable {
+std::function<void()> make_task(array arr, bool signal) {
+  auto task = [arr = std::move(arr), signal]() mutable {
     auto pool = new_scoped_memory_pool();
-    for (auto& d : deps) {
-      d.wait();
-    }
     auto s = arr.primitive().stream();
     auto command_buffer = increment_command_buffer(s);
+    for (auto& i : arr.inputs()) {
+      if (i.event().valid() && i.event().stream() != arr.primitive().stream()) {
+        // TODO, consider committing the buffer and encoding a wait in the new
+        // buffer rather than on the task thread
+        i.event().wait();
```

Comment:
This is quite interesting as it would allow us to continue encoding things in the GPU that will run immediately when the CPU signals the dependencies.

Otoh if the CPU takes too much time the GPU may run too far ahead taking up all the memory and so on. I think this is a good first choice but I see us implementing a compromise in the future. Something like the `malloc_or_wait`...